src/compute-client/src/plan/mod.rs

```diff
@@ -1674,10 +1689,53 @@ This is not expected to cause incorrect results, but could indicate a performanc
                 source.arguments.operators = Some(mfp);
             }
         }
+    }
 
-        mz_repr::explain::trace_plan(&dataflow);
+    /// Refines the plans of objects to be built as part of `dataflow` to take advantage
+    /// of monotonic operators if the dataflow refers to a single-time, i.e., is for a
+    /// one-shot SELECT query.
+    fn refine_single_time_dataflow(dataflow: &mut DataflowDescription<Self>) {
+        // Check if we have a one-shot SELECT query, i.e., a single-time dataflow.
+        if !dataflow.is_single_time() {
+            return;
+        }
 
-        Ok(dataflow)
+        // Upgrade single-time plans to monotonic.
+        for build_desc in dataflow.objects_to_build.iter_mut() {
+            let mut todo = vec![&mut build_desc.plan];
+            while let Some(expression) = todo.pop() {
+                match expression {
+                    Plan::Reduce { plan, .. } => {
+                        use crate::plan::reduce::CollationPlan;
+                        // Upgrade non-monotonic hierarchical plans to monotonic with mandatory consolidation.
+                        match plan {
+                            ReducePlan::Collation(CollationPlan { hierarchical, .. }) => {
+                                hierarchical.as_mut().map(|plan| plan.flag_snapshot());
```

Comment:
`flag_snapshot` is referring to the pre-precondition under which we apply the action and not the action itself.

Can we call these methods `as_monotonic(must_consolidate: true)` instead? I see that one of the two variants already has `into_monotonic`. I'm fine with variations of this as long as we agree on some more descriptive and consistent API across the board.

Alternatively, you can just inline the `flag_snapshot` logic for `ReducePlan` and `TopK` in the `refine_single_time_dataflow` body (I'm not sure what's the use of this method except reducing readability clutter).

Reply:
Yes, happy to change the method to represent better the action being applied. I followed your suggestion of `as_monotonic` with `must_consolidate` as a parameter, and thus made the implementations also uniformly set `must_consolidate` appropriately if the plan was already monotonic. This change is done in 9317df235aa652293707e74e2887534f3bdf77c1.